WindowsAzureMessaging/WindowsAzureMessaging/SBNotificationHub.m

```diff
@@ -378,19 +378,58 @@ - (void)deleteRegistrationWithName:(NSString *)templateName completion:(void (^)
 - (void)unregisterAllWithDeviceToken:(NSData *)deviceTokenData completion:(void (^)(NSError *))completion {
     if (deviceTokenData == nil) {
         if (completion) {
-            completion([SBNotificationHubHelper errorForNullDeviceToken]);
+            dispatch_async(dispatch_get_main_queue(), ^{
+                completion([SBNotificationHubHelper errorForNullDeviceToken]);
+            });
         }
 
         return;
     }
 
     NSString *deviceToken = [self convertDeviceToken:deviceTokenData];
-
-    SBThreadParameter *parameter = [[SBThreadParameter alloc] init];
-    parameter.deviceToken = deviceToken;
-    parameter.completion = completion;
-    parameter.isMainThread = [[NSThread currentThread] isMainThread];
-    [self performSelectorInBackground:@selector(deleteAllRegistrationThread:) withObject:parameter];
+    
+    [self retrieveAllRegistrationsWithDeviceToken:deviceToken completion:^(NSArray *registrations, NSError *error) {
+        
+        if (error) {
```

Comment:
What about checking if error && completion, that way there is no need to submit the block to the main queue if there is no completion to be executed.